// adonis-captcha-guard/build/recaptcha-XJ375K4E.js
import {
  E_CAPTCHA_MISSING_TOKEN
} from "./chunk-NKLW5P6S.js";
import {
  Captcha
} from "./chunk-RSIWGLWZ.js";

// src/captchas/recaptcha.ts
import got from "got";
var RecaptchaService = class extends Captcha {
  constructor(ctx, config) {
    super(ctx, config);
    this.config = config;
  }
  verifyUrl = "https://www.google.com/recaptcha/api/siteverify";
  tokenParamName = "token";
  async validate() {
    if (!this.hasToken()) {
      throw new E_CAPTCHA_MISSING_TOKEN([this.tokenParamName]);
    }
    console.log(this.config.secret);
    const token = this.getToken();
    const body = await got.post(
      this.verifyUrl,
      {
        form: {
          secret: this.config.secret,
          response: token
        }
      }
    ).text();
    const decodeBody = JSON.parse(body);
    return {
      success: Boolean(decodeBody.success),
      challengeTimestamp: decodeBody.challenge_ts,
      hostname: decodeBody.hostname,
      errorCodes: decodeBody["error-codes"]
    };
  }
};
export {
  RecaptchaService
};
