// @adonisjs/eslint-plugin/build/index.js
// src/rules/prefer_lazy_listener_import.ts
import { AST_NODE_TYPES } from "@typescript-eslint/utils";

// src/utils.ts
import { ESLintUtils } from "@typescript-eslint/utils";
var createEslintRule = ESLintUtils.RuleCreator((ruleName) => `https://github.com/adonisjs/eslint-plugin-adonisjs#${ruleName}`);

// src/rules/prefer_lazy_listener_import.ts
var prefer_lazy_listener_import_default = createEslintRule({
  name: "prefer-lazy-listener-import",
  defaultOptions: [],
  meta: {
    type: "problem",
    fixable: "code",
    docs: {
      description: "(Needed for HMR) Prefer lazy listener import over standard import"
    },
    schema: [],
    messages: {
      preferLazyListenerImport: "Replace standard import with lazy listener import"
    }
  },
  create: function(context) {
    const importNodes = {};
    const importIdentifiers = [];
    let emitterIdentifier = "";
    function isEmitterOnCallExpression(node, routerIdentifier) {
      return node.callee.type === AST_NODE_TYPES.MemberExpression && node.callee.object.type === AST_NODE_TYPES.Identifier && node.callee.object.name === routerIdentifier && node.callee.property.type === AST_NODE_TYPES.Identifier && node.callee.property.name === "on";
    }
    return {
      /**
       * Track all imported identifiers
       * Also get the local name of the emitter import
       */
      ImportDeclaration(node) {
        for (const specifier of node.specifiers) {
          if (specifier.type === "ImportDefaultSpecifier" || specifier.type === "ImportSpecifier") {
            importIdentifiers.push(specifier.local.name);
            importNodes[specifier.local.name] = node;
          }
        }
        if (node.source.value === "@adonisjs/core/services/emitter") {
          if (node.specifiers[0] && node.specifiers[0].type === "ImportDefaultSpecifier") {
            emitterIdentifier = node.specifiers[0].local.name;
          }
        }
      },
      CallExpression(node) {
        if (!isEmitterOnCallExpression(node, emitterIdentifier)) {
          return;
        }
        const secondArgument = node.arguments[1];
        if (secondArgument.type !== AST_NODE_TYPES.ArrayExpression) {
          return;
        }
        for (const element of secondArgument.elements) {
          if (!element) {
            continue;
          }
          if (element.type !== "Identifier" || !importIdentifiers.includes(element.name)) {
            continue;
          }
          context.report({
            node: importNodes[element.name],
            messageId: "preferLazyListenerImport",
            fix(fixer) {
              const importPath = importNodes[element.name].source.raw;
              const newImportDeclaration = `const ${element.name} = () => import(${importPath})`;
              return fixer.replaceText(importNodes[element.name], newImportDeclaration);
            }
          });
        }
      }
    };
  }
});

// src/rules/prefer_lazy_controller_import.ts
import { AST_NODE_TYPES as AST_NODE_TYPES2 } from "@typescript-eslint/utils";
var HTTP_METHODS = ["get", "post", "put", "delete", "patch"];
var prefer_lazy_controller_import_default = createEslintRule({
  name: "prefer-lazy-controller-import",
  defaultOptions: [],
  meta: {
    type: "problem",
    fixable: "code",
    docs: {
      description: "(Needed for HMR) Prefer lazy controller import over standard import"
    },
    schema: [],
    messages: {
      preferLazyControllerImport: "Replace standard import with lazy controller import"
    }
  },
  create: function(context) {
    const importNodes = {};
    const importIdentifiers = [];
    let routerIdentifier = "";
    function isRouteCallExpression(node, identifier) {
      return node.callee.type === AST_NODE_TYPES2.MemberExpression && node.callee.object.type === AST_NODE_TYPES2.Identifier && node.callee.object.name === identifier && node.callee.property.type === AST_NODE_TYPES2.Identifier && HTTP_METHODS.includes(node.callee.property.name);
    }
    function isRouteResourceCallExpression(node, identifier) {
      return node.callee.type === AST_NODE_TYPES2.MemberExpression && node.callee.object.type === AST_NODE_TYPES2.Identifier && node.callee.object.name === identifier && node.callee.property.type === AST_NODE_TYPES2.Identifier && node.callee.property.name === "resource";
    }
    return {
      /**
       * Track all imported identifiers
       * Also get the local name of the router import
       */
      ImportDeclaration(node) {
        for (const specifier of node.specifiers) {
          if (specifier.type === "ImportDefaultSpecifier" || specifier.type === "ImportSpecifier") {
            importIdentifiers.push(specifier.local.name);
            importNodes[specifier.local.name] = node;
          }
        }
        if (node.source.value === "@adonisjs/core/services/router") {
          if (node.specifiers[0] && node.specifiers[0].type === "ImportDefaultSpecifier") {
            routerIdentifier = node.specifiers[0].local.name;
          }
        }
      },
      CallExpression(node) {
        let controller = null;
        if (isRouteCallExpression(node, routerIdentifier)) {
          const secondArgument = node.arguments[1];
          if (secondArgument.type === AST_NODE_TYPES2.ArrayExpression) {
            controller = secondArgument.elements[0];
          }
        }
        if (isRouteResourceCallExpression(node, routerIdentifier)) {
          controller = node.arguments[1];
        }
        if (!controller) {
          return;
        }
        if (controller.type !== "Identifier" || !importIdentifiers.includes(controller.name)) {
          return;
        }
        context.report({
          node: importNodes[controller.name],
          messageId: "preferLazyControllerImport",
          fix(fixer) {
            const importPath = importNodes[controller.name].source.raw;
            const newImportDeclaration = `const ${controller.name} = () => import(${importPath})`;
            return fixer.replaceText(importNodes[controller.name], newImportDeclaration);
          }
        });
      }
    };
  }
});

// index.ts
var index_default = {
  rules: {
    "prefer-lazy-controller-import": prefer_lazy_controller_import_default,
    "prefer-lazy-listener-import": prefer_lazy_listener_import_default
  }
};
export {
  index_default as default
};
